src/transformers/modeling_utils.py

```diff
@@ -515,6 +515,64 @@ def set_initialized_submodules(model, state_dict_keys):
             module._is_hf_initialized = True
 
 
+def _end_ptr(tensor: torch.Tensor) -> int:
+    if tensor.nelement():
+        stop = tensor.view(-1)[-1].data_ptr() + tensor.element_size()
+    else:
+        stop = tensor.data_ptr()
+    return stop
+
+
+def _find_disjoint(tensors: List[Set[str]], state_dict: Dict[str, torch.Tensor]) -> Tuple[List[Set[str]], Set[str]]:
+    filtered_tensors = []
+    for shared in tensors:
+        if len(shared) < 2:
+            filtered_tensors.append(shared)
+            continue
+
+        areas = []
+        for name in shared:
+            tensor = state_dict[name]
+            areas.append((tensor.data_ptr(), _end_ptr(tensor), name))
+        areas.sort()
+
+        _, last_stop, last_name = areas[0]
+        filtered_tensors.append({last_name})
+        for start, stop, name in areas[1:]:
+            if start >= last_stop:
+                filtered_tensors.append({name})
+            else:
+                filtered_tensors[-1].add(name)
+            last_stop = stop
+    disjoint_tensors = []
+    shared_tensors = []
+    for tensors in filtered_tensors:
+        if len(tensors) == 1:
+            disjoint_tensors.append(tensors.pop())
+        else:
+            shared_tensors.append(tensors)
+    return shared_tensors, disjoint_tensors
+
+
+def _find_identical(tensors: List[Set[str]], state_dict: Dict[str, torch.Tensor]) -> Tuple[List[Set[str]], Set[str]]:
```

Comment:
This does not seem tested by the small dummy test (but is most probably tested by some models that do have this?)

Reply:
Yes, because `area` contains the device as a key.

Nothing in transformers could trigger this, only `optimum` did trigger this. It requires some low level manipulation of storages and tensors.

Reply:
Identical was already previously tested by other tests. (Since it's a regular flow for shared tensors with `tie_weights`.)

The dummy test only tests the new path for disjoint tensors (which cannot be triggered by regular transformers code)